src/paths/bellman_ford.c

```diff
@@ -203,3 +210,284 @@ int igraph_shortest_paths_bellman_ford(const igraph_t *graph,
+
+        for (i = 0; i < no_of_nodes; i++) {
+            if (i == from) {
+                /* i is the start vertex */
+                VECTOR(*predecessors)[i] = i;
+            } else if (parents[i] <= 0) {
+                /* i was not reached */
+                VECTOR(*predecessors)[i] = -1;
+            } else {
+                /* i was reached via the edge with ID = parents[i] - 1 */
+                VECTOR(*predecessors)[i] = IGRAPH_OTHER(graph, parents[i] - 1, i);
+            }
+        }
+    }
+
+    /* Create `inbound_edges' if needed */
+    if (inbound_edges) {
+        IGRAPH_CHECK(igraph_vector_long_resize(inbound_edges, no_of_nodes));
+
+        for (i = 0; i < no_of_nodes; i++) {
+            if (parents[i] <= 0) {
+                /* i was not reached */
+                VECTOR(*inbound_edges)[i] = -1;
+            } else {
+                /* i was reached via the edge with ID = parents[i] - 1 */
+                VECTOR(*inbound_edges)[i] = parents[i] - 1;
+            }
+        }
+    }
+
+    /* Reconstruct the shortest paths based on vertex and/or edge IDs */
+    if (vertices || edges) {
+        for (IGRAPH_VIT_RESET(tovit), i = 0; !IGRAPH_VIT_END(tovit); IGRAPH_VIT_NEXT(tovit), i++) {
+            long int node = IGRAPH_VIT_GET(tovit);
+            long int size, act, edge;
+            igraph_vector_t *vvec = 0, *evec = 0;
+            if (vertices) {
+                vvec = VECTOR(*vertices)[i];
+                igraph_vector_clear(vvec);
+            }
+            if (edges) {
+                evec = VECTOR(*edges)[i];
+                igraph_vector_clear(evec);
+            }
+
+            IGRAPH_ALLOW_INTERRUPTION();
+
+            size = 0;
+            act = node;
+            while (parents[act]) {
+                size++;
+                edge = parents[act] - 1;
+                act = IGRAPH_OTHER(graph, edge, act);
+            }
+            if (vvec) {
+                IGRAPH_CHECK(igraph_vector_resize(vvec, size + 1));
+                VECTOR(*vvec)[size] = node;
+            }
+            if (evec) {
+                IGRAPH_CHECK(igraph_vector_resize(evec, size));
+            }
+            act = node;
+            while (parents[act]) {
+                edge = parents[act] - 1;
+                act = IGRAPH_OTHER(graph, edge, act);
+                size--;
+                if (vvec) {
+                    VECTOR(*vvec)[size] = act;
+                }
+                if (evec) {
+                    VECTOR(*evec)[size] = edge;
+                }
+            }
+        }
+    }
+
+    igraph_vector_destroy(&dist);
+    IGRAPH_FINALLY_CLEAN(1);
+
+    if (!all_to) {
+        igraph_vit_destroy(&tovit);
+        IGRAPH_FINALLY_CLEAN(1);
+    }
+    free(parents);
+    igraph_dqueue_destroy(&Q);
+    igraph_vector_destroy(&clean_vertices);
+    igraph_vector_destroy(&num_queued);
+    igraph_lazy_inclist_destroy(&inclist);
+    IGRAPH_FINALLY_CLEAN(4);
```

Comment:
Now we _do_ have 5 things to clean, including the `igraph_Free(parents)`, so this should indeed be 5 now.